# SQL Project

In this project we will use a database containing data on books, publishers, authors, customer ratings and reviews of books to look at different aspects of the dataset.

<a class="anchor" id="toc"></a>

### Table of Contents

#### 1. [Importing Libraries](#chapter1)

#### 2. [Goals of the study](#chapter2)

#### 3. [Study the tables](#chapter3)
* ['books' table](#section1)
* ['authors' table](#section2)
* ['publishers' table](#section3)
* ['ratings' table](#section4)
* ['reviews' table](#section5)

#### 4. [Tasks](#chapter4)
* [Find the number of books released after January 1, 2000](#section6)
* [Find the number of user reviews and the average rating for each book](#section7)
* [Identify the publisher that has released the greatest number of books with more than 50 pages](#section8)
* [Identify the author with the highest average book rating: look only at books with at least 50 ratings](#section9)
* [Find the average number of text reviews among users who rated more than 50 books](#section10)

<a class="anchor" id="chapter1"></a>

## Import libraries

In [1]:
# import libraries
import pandas as pd
from sqlalchemy import create_engine


db_config = {'user': 'praktikum_student',         # user name
             'pwd': 'Sdf4$2;d-d30pp', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # the name of the data base

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

[Back to Table of Contents](#toc)

<a class="anchor" id="chapter2"></a>

## Goals of the study

The goal of this study is to use a dataset provided from one of the services competing in the book apps market to generate a value proposition for a new product.

[Back to Table of Contents](#toc)

<a class="anchor" id="chapter3"></a>

## Study the tables

In [2]:
# function that takes a query and return dataframe for general use
def queryResult(q):
    return pd.io.sql.read_sql(q, con = engine)

<a class="anchor" id="section1"></a>

### `books` table:

Contains data on books:
- `book_id`
- `author_id`
- `title`
- `num_pages` — number of pages
- `publication_date`
- `publisher_id`

In [3]:
query='''
        SELECT
            *
        FROM
            books
        '''
books=queryResult(query)

In [4]:
books.head(10)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

<a class="anchor" id="section2"></a>

### `authors` table:

Contains data on authors:

- `author_id`
- `author`

In [5]:
query='''
        SELECT
            *
        FROM
            authors
        '''
authors=queryResult(query)

In [6]:
authors.head(10)

author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts

<a class="anchor" id="section3"></a>

### `publishers` table:

Contains data on publishers:

- `publisher_id`
- `publisher`

In [7]:
query='''
        SELECT
            *
        FROM
            publishers
        '''
publishers=queryResult(query)

In [8]:
publishers.head(10)

publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers

<a class="anchor" id="section4"></a>

### `ratings` table:

Contains data on user ratings:

- `rating_id`
- `book_id`
- `username` — the name of the user who rated the book
- `rating`

In [9]:
query='''
        SELECT
            *
        FROM
            ratings
        '''
ratings=queryResult(query)

In [10]:
ratings.head(10)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

<a class="anchor" id="section5"></a>

### `reviews` table:

Contains data on customer reviews:

- `review_id`
- `book_id`
- `username` — the name of the user who reviewed the book
- `text` — the text of the review

In [11]:
query='''
        SELECT
            *
        FROM
            reviews
        '''
reviews=queryResult(query)

In [12]:
reviews.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

[Back to Table of Contents](#toc)

<a class="anchor" id="chapter4"></a>

## Tasks

- Find the number of books released after January 1, 2000.
- Find the number of user reviews and the average rating for each book.
- Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).
- Identify the author with the highest average book rating (look only at books with at least 50 ratings).
- Find the average number of text reviews among users who rated more than 50 books.

<a class="anchor" id="section6"></a>

***1. Find the number of books released after January 1, 2000***

In [13]:
query='''
SELECT
    COUNT("book_id") AS total_books
FROM
    books
WHERE
    "publication_date"::date > '2000-01-01'
'''
queryResult(query)

total_books
0          819

There are a total of 819 books that were released after January 1, 2000.

<a class="anchor" id="section7"></a>

***2. Find the number of user reviews and the average rating for each book.***

In [14]:
query='''
SELECT
    books."book_id",
    books."title",
    COUNT(reviews."review_id") AS num_of_reviews,
    AVG(ratings."rating") AS avg_rating
FROM
    reviews
    RIGHT JOIN books ON reviews."book_id" = books."book_id"
    RIGHT JOIN ratings ON books."book_id" = ratings."book_id"
    
GROUP BY
    books."title",books."book_id"
ORDER BY
    num_of_reviews DESC;
    
'''


queryResult(query)

book_id                                              title  \
0        948                            Twilight (Twilight  #1)   
1        750                The Hobbit  or There and Back Again   
2        673                             The Catcher in the Rye   
3        302  Harry Potter and the Prisoner of Azkaban (Harr...   
4        299  Harry Potter and the Chamber of Secrets (Harry...   
..       ...                                                ...   
995      221                          Essential Tales and Poems   
996      808                            The Natural Way to Draw   
997      191  Disney's Beauty and the Beast (A Little Golden...   
998       83    Anne Rice's The Vampire Lestat: A Graphic Novel   
999      672   The Cat in the Hat and Other Dr. Seuss Favorites   

     num_of_reviews  avg_rating  
0              1120    3.662500  
1               528    4.125000  
2               516    3.825581  
3               492    4.414634  
4               480    4.287500  
..              ...         ...  
995               0    4.000000  
996               0    3.000000  
997               0    4.000000  
998               0    3.666667  
999               0    5.000000  

[1000 rows x 4 columns]

The book with an average rating of 5 with the largest number of reviews is "A Dirty Job (Grim Reaper #1)"
The book with the lowest average rating in the list is "Harvesting the Heart"

<a class="anchor" id="section8"></a>

***3. Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).***

In [15]:
query='''
SELECT
    publishers."publisher",
    COUNT(books."book_id") AS num_of_books
FROM
    books
    INNER JOIN publishers ON books."publisher_id" = publishers."publisher_id"
WHERE
    books."num_pages" > 50
GROUP BY
    publishers."publisher"
ORDER BY
    num_of_books DESC
LIMIT(1)
'''
queryResult(query)

publisher  num_of_books
0  Penguin Books            42

The publisher that has released the greatest number of books with more than 50 pages is "Penguin Books" with 42 total books.

<a class="anchor" id="section9"></a>

***4. Identify the author with the highest average book rating (look only at books with at least 50 ratings).***

In [16]:
query='''
SELECT
    books."author_id",
    authors."author",
    AVG(ratings."rating") AS avg_rating
FROM
    authors RIGHT JOIN books 
        ON authors."author_id" = books."author_id"
    RIGHT JOIN ratings 
        ON books."book_id" = ratings."book_id"
GROUP BY
    books."author_id", authors."author"
HAVING
    COUNT(ratings."rating_id") > 50
ORDER BY
    avg_rating DESC
LIMIT (1)
'''
queryResult(query)



author_id                      author  avg_rating
0        236  J.K. Rowling/Mary GrandPré    4.288462

J.K. Rowling/Mary GrandPré is the author with the highest average book rating of 4.3 (that has at least 50 ratings).

<a class="anchor" id="section10"></a>

***5. Find the average number of text reviews among users who rated more than 50 books.***

In [17]:
query='''
SELECT
    AVG(text_cnt) AS text_avg
    FROM(
        SELECT 
            t_M_rate.username,
            rate_cnt,
            count(text) AS text_cnt
        FROM (
                (SELECT
                    *      
                FROM (
                    SELECT
                        ratings."username",
                        COUNT(ratings."username") AS rate_cnt
                    FROM ratings 
                    GROUP BY ratings."username"
                    ) AS t_rate
                WHERE
                    rate_cnt>50
                ) AS t_M_rate
            left join 
                (SELECT 
                    reviews."username",
                    reviews.text
                FROM reviews
                ) as t_M_reviews
            ON t_M_rate."username" = t_M_reviews."username"
        )
        GROUP BY t_M_rate.username , rate_cnt
    ) AS T
'''
queryResult(query)

text_avg
0  24.333333

The average number of text reviews among users who rated more than 50 books is 24.333

[Back to Table of Contents](#toc)